base/task.jl

```diff
@@ -365,6 +365,67 @@ function wait(t::Task)
     nothing
 end
 
+# Wait multiple tasks
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks; all=false, failfast=false)
+    foreach(waiting_tasks) do @nospecialize t
+        if !(t isa Task)
+            error("Expected an iterator of `Task` object")
+        end
+    end
+
+    chan = Channel{Tuple{Int,Task}}(Inf)
+    tasks = Task[]
+    done_mask = Bool[]
+    exception = false
+    nremaining::Int = 0
+
+    for (i, t) in enumerate(waiting_tasks)
+        t = t::Task
+        push!(tasks, t)
+        if istaskdone(t)
+            push!(done_mask, true)
+            exception |= istaskfailed(t)
+        else
+            nremaining += 1
+            push!(done_mask, false)
+            schedule(Task(() -> begin
+                _wait(t)
+                put!(chan, (i, t))
+            end))
+        end
+    end
+
+    if nremaining == 0
+        close(chan)
+        return tasks, Task[]
+    elseif any(done_mask) && (!all || (failfast && exception))
+        close(chan)
+        return tasks[done_mask], tasks[.~done_mask]
```

Comment:
It looks like the `any` option can leak a non-trivial amount of resources on each call. There is not a particularly good API for avoiding that either. It needs to go back through all of the `_wait2` tasks and delete them from the donenotify list if they are present there, using something like:
```julia
@lock t.donenotify Base.list_deletefirst!((t.donenotify::ThreadSynchronizer).waitq, t2)
```

Reply:
@vtjnash Thank you for your advice. I'll insert the cleanup code.